File: AlgorithmicToolbox/week3_greedy_algorithms/7_maximum_salary/7_maximum_salary.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# python3
"""
As the last question of a successful interview, your boss gives you a few pieces of paper
with numbers on it and asks you to compose a largest number from these numbers. The
resulting number is going to be your salary, so you are very much interested in maximizing
this number. How can you do this?
"""
from itertools import permutations


def largest_number_naive(numbers):
    numbers = list(map(str, numbers))

    largest = 0

    for permutation in permutations(numbers):
        largest = max(largest, int("".join(permutation)))

    return largest


def largest_number(numbers):
    numbers = list(map(str, numbers))
    print(numbers)
    salary = []
    while len(numbers) > 0:
        max_digit = 0
        for number in numbers:
            if int(str(number)+str(max_digit))>=int(str(max_digit)+str(number)):
                max_digit = number
        salary.append(max_digit)
        numbers.remove(max_digit)
    return int("".join(salary))

if __name__ == '__main__':
    n = int(input())
    input_numbers = input().split()
    assert len(input_numbers) == n
    print(largest_number(input_numbers))

File: AlgorithmicToolbox/week3_greedy_algorithms/7_maximum_salary/7_maximum_salary.cpp
#include <algorithm>
#include <sstream>
#include <iostream>
#include <vector>
#include <string>

/*
As the last question of a successful interview, your boss gives you a few pieces of 
paper with numbers on it and asks you to compose a largest number from these numbers.
The resulting number is going to be your salary, so you are very much interested in
maximizing this number. How can you do this?
*/

std::vector<std::string> largest_number(std::vector<std::string> a) {
    std::vector<std::string> salary;
    while (a.size() > 0) {
        std::string max_digit{'0'};
        int index{0};
        for (int i = 0; i < a.size(); ++i) {
            if (std::stoi(a[i] + max_digit) >= std::stoi(max_digit + a[i])) {
                max_digit = a[i];
                index = i;
            }
            
        }
        salary.push_back(max_digit);
        a.erase(a.begin() + index);
    }

    return salary;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<std::string> a(n);
    for (size_t i = 0; i < a.size(); i++) {
        std::cin >> a[i];
    }

    std::vector<std::string> salary = largest_number(a);
    for (auto x : salary) {
        std::cout << x;
    }

    return 0;
}
